patches/minecraft/net/minecraft/server/network/ServerLoginPacketListenerImpl.java.patch

```diff
@@ -15,6 +15,21 @@
           this.m_10055_();
        } else if (this.f_10019_ == ServerLoginPacketListenerImpl.State.DELAY_ACCEPT) {
           ServerPlayer serverplayer = this.f_10018_.m_6846_().m_11259_(this.f_10021_.getId());
+@@ -116,6 +_,14 @@
+                this.f_10019_ = ServerLoginPacketListenerImpl.State.DELAY_ACCEPT;
+                this.f_10024_ = serverplayer1;
+             } else {
++               Component reason = net.minecraftforge.event.ForgeEventFactory.firePrePlayerLoggedIn(serverplayer1, this.f_10013_);
++               if (reason != null) {
++                  this.f_10013_.m_129514_(new ClientboundDisconnectPacket(reason), (p_9828_) -> this.f_10013_.m_129507_(reason));
++                  this.f_10013_.m_129540_();
++                  this.f_10018_.m_18709_(this.f_10013_::m_129541_);
++                  return;
++               }
```

Comment:
So I put it here because as I understand it the DELAY_ACCEPT only happens when the user with said UUID is already logged in and it's delaying to see if they're just an artifact of lag (and if they're not the connection gets denied for "You're already logged in"). 

I didn't add it to PlayerList.canPlayerLogin as that's where the whitelist is checked. I suppose it could go in there. However, at that point you don't have access to the player object so I figured maybe it'd be more useful for mods utilizing this API to have such access.

Reply:
Moving the logic to the hook. And when using the implementation from `ServerLoginPacketListenerImpl#disconnect` I get [this](https://i.gyazo.com/78681e94197d139d2c63dfdd9c2e3508.png) disconnect error. However, when using the one above it disconnects with the correct message. I attempted to try find out why and I came to the conclusion that the logic was too late for this packet to be sent (?). Not 100% certain on that explanation but it's something about that specific packet. And I also did attempt to change the text in the kick message but regardless of the length of the disconnect message it always resulted in the same buffer error.